[MRG+2] Implement two non-uniform strategies for KBinsDiscretizer (discrete branch)

TomDLT opened the pull request:
Fixes #9338.

This PR implements two non-uniform strategies for KBinsDiscretizer (in discrete branch #9342), adding a parameter `strategy`:
- `KBinsDiscretizer(strategy='uniform')`: previous behavior, constant width bins.
- `KBinsDiscretizer(strategy='quantile')`: new behavior (default), based on `np.percentile`.
- `KBinsDiscretizer(strategy='kmeans')`: new behavior, based on `KMeans`.

It adds the following example:
![figure_1](https://user-images.githubusercontent.com/11065596/41465630-3534fa88-709f-11e8-96bc-19541342a877.png)

TomDLT committed:
Implement two new strategies for KBinsDiscretizer

TomDLT committed:
FIX python viridis import

jnothman commented:
Nice!​

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -52,6 +56,22 @@ class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+        kmeans
+            Widths are defined by a k-means on each features.
+
+    random_state : int, RandomState instance or None (default)
```
Seeing as we're doing it in 1d, can't we just initialise kmeans at the midpoints of a uniform split, and make it deterministic?

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -152,16 +181,45 @@ def fit(self, X, y=None):
+                    n_clusters = self.n_bins_[jj]
+                    km = KMeans(n_clusters=n_clusters,
+                                random_state=self.random_state).fit(col)
+                    centers = np.sort(km.cluster_centers_[:, 0],
```
I don't think we'd need this sort if we initialised with linspace

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -152,16 +181,45 @@ def fit(self, X, y=None):
+
+                if self.strategy == 'quantile':
+                    n_quantiles = self.n_bins_[jj] + 1
+                    qt = QuantileTransformer(n_quantiles=n_quantiles).fit(col)
```
This seems overkill. Just use `boundaries = np.percentile(col, n_quantiles + 1)`

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -142,6 +166,11 @@ def fit(self, X, y=None):
+        valid_strategy = ('uniform', 'quantile', 'kmeans')
+        if self.strategy not in valid_strategy:
+            raise ValueError("Valid options for 'strategy' are {}. "
+                             "Got 'strategy = {}' instead."
```
use {!r} for quoting

jnothman commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -228,3 +231,75 @@ def test_one_hot_encode_with_ignored_features():
+def test_nonuniform_strategies():
+    X = np.array([0, 1, 2, 3, 9, 10]).reshape(-1, 1)
+
+    # with 2 bins
```
Please also test the size of `bin_widths_`

jnothman reviewed (commented):
I think we should treat this as if it'll be released with KBinsDiscretizer. We should set the default strategy to 'quantile'. We should consider API changes to make this more consistent. We should update the what's new message rather than adding one.

I was a bit surprised to see you've only added, and not modified tests. But the only thing I'm missing there is checking the consistency of 1-column and multi-column transformations.

TomDLT committed:
ENH store edges instead of widths

TomDLT committed:
ENH change default to "quantile" and make "kmeans" deterministic

TomDLT committed:
ENH improve tests and error messages

TomDLT commented:
Thanks for the review.
- I changed to storing the edges in all cases, instead of widths. It avoids successive `diff` and `cumsum` in non-uniform strategies, and makes the API more consistent across strategies.
- I removed the offset attribute, since the information is now in the edges.
- I changed the default to 'quantile'.
- I made 'kmeans' strategy deterministic with a uniform init.

TomDLT committed:
CLN simplify branching

jnothman commented:
> I changed to storing the edges in all cases, instead of widths. It avoids successive diff and cumsum in non-uniform strategies, and makes the API more consistent across strategies.

And I suppose we're not worried about a slow-down in the uniform case?

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -265,24 +305,20 @@ def _transform(self, X):
-            X /= bin_width
+        bin_edges = self.bin_edges_[trans]
+        for jj in range(X.shape[1]):
+            # Values which are a multiple of the bin width are susceptible to
```
"a multiple of the bin width" -> "close to the bin edge"

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -265,24 +305,20 @@ def _transform(self, X):
+        for jj in range(X.shape[1]):
+            # Values which are a multiple of the bin width are susceptible to
+            # numeric instability. Add eps to X so these values are binned
+            # correctly. See documentation for numpy.isclose for an explanation
```
Maybe "correctly" -> "correctly with respect to their decimal truncation"???

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -140,28 +148,60 @@ def fit(self, X, y=None):
+                edges = np.r_[col.min(), edges, col.max()]
+
+            if edges[0] == edges[-1] and self.n_bins_[jj] > 2:
+                warnings.warn("Features %d is constant and will be "
```
Features -> Feature

TomDLT commented:
Good point, if you think it's important, I can add a branch in `_transform` to be faster on the uniform strategy.

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -52,19 +55,26 @@ class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+            Widths are defined by a quantile transform, to have a uniform
+            number of samples in each bin.
+        kmeans
+            Widths are defined by a k-means on each features.
```
features -> feature

Perhaps:
"Values in each bin have the same nearest center of a k-means cluster."

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -52,19 +55,26 @@ class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+        uniform
+            All bins in each feature have identical widths.
+        quantile
+            Widths are defined by a quantile transform, to have a uniform
```
"All bins have the same number of points"

jnothman reviewed (approved):
Great work. All nitpicks. LGTM!

jnothman commented:
> Good point, if you think it's important, I can add a branch in _transform to be faster on the uniform strategy.

I don't yet have reason to believe it's important... But yes, we could consider that now or in the future.

TomDLT committed:
FIX example

TomDLT committed:
FIX compatibility with numpy 1.8

TomDLT committed:
CLN just renaming variables

jnothman commented:
I think once this is merged, we should move the discretizer to `_encoders.py`. This would resolve conflicts with master.

jnothman commented:
Would @qinhanmin2014 like to give this a second look?

qinhanmin2014 commented:
I promise to give a review next weekend, and a second review from someone else before next weekend is welcomed. Sorry for the delay but I'm surrounded by school work at the end of the semester.

jnothman commented:
That's fine! Just thought you might be a good person to give it a go if
available.​

qinhanmin2014 commented on sklearn/preprocessing/discretization.py:
```diff
@@ -140,28 +150,61 @@ def fit(self, X, y=None):
+                edges = np.asarray(np.percentile(col, quantiles))
+
+            elif self.strategy == 'kmeans':
+                from ..cluster import KMeans
```
Any specific reason to import here?

TomDLT replied:
Importing loop problem, but it seems not to happen anymore so I moved it to the top.

qinhanmin2014 commented on doc/modules/preprocessing.rst:
```diff
@@ -565,7 +563,7 @@ example, these intervals are defined as:
 
   >>> est.transform(X)                      # doctest: +SKIP
-  array([[ 0., 2., 1.],
+  array([[ 0., 1., 1.],
```
If I understand correctly, the result change because we're now using ``strategy='quantile'``, right?
If so, I think we need to modify the intervals above, otherwise they're not consistent:
```
  - feature 1: :math:`{[-\infty, 0), [0, 3), [3, \infty)}`
  - feature 2: :math:`{[-\infty, 4), [4, 5), [5, \infty)}`
  - feature 3: :math:`{[-\infty, 13), [13, \infty)}`
```

TomDLT replied:
Good catch !

qinhanmin2014 reviewed (commented):
The code LGTM. Will look at the examples and tests in a couple of hours.

qinhanmin2014 commented on sklearn/preprocessing/discretization.py:
```diff
@@ -8,12 +11,14 @@
+from ..utils.fixes import np_version
 
 
 class KBinsDiscretizer(BaseEstimator, TransformerMixin):
```
Need to update the doc below:
``Bins continuous data into k equal width intervals.`` is no longer the case I think.

TomDLT replied:
Thanks

qinhanmin2014 commented:
@jnothman @TomDLT  Will it be better to change the default value of ``n_bins`` to e.g., 3 or 5. ``n_bins=2`` seems strange from my side, even though I'm aware that we're going to provide ways to automatically determine the value of ``n_bins``

qinhanmin2014 commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -215,16 +230,76 @@ def test_encode_options():
+    assert_array_equal(expected_3bins, Xt.ravel())
+
+
+def test_kmeans_strategy():
```
Could you explain the purpose of the test (maybe a comment)? It seems a bit strange (though I understand how it pass) and honestly seems redundant from my side.

TomDLT replied:
It is indeed redundant so I removed it.

jnothman commented:
Without looking at the code, 'auto' in `np.histogram` is:
```
q25, q75 = np.percentile(X, [25, 75], axis=0)
iqr = q75 - q25
fd = 2 * iqr / X.shape[0] ** (1/3)
n_bins_fd = np.ceil(np.ptp(X, axis=0) / fd)
n_bins_sturges = np.ceil(np.log2(X.shape[0])) + 1
n_bins = np.maximum(n_bins_fd, n_bins_sturges).astype(int)
```
so if we want a better default, it's not hard to implement....

qinhanmin2014 commented on examples/preprocessing/plot_discretization_strategies.py:
```diff
@@ -0,0 +1,94 @@
+==========================================================
+
+This example presents the different strategies implemented in KBinsDiscretizer:
+- 'uniform': The discretization is uniform in each feature, which means that
```
Formatting issue according to Circle. I guess a blank line will solve the problem.

qinhanmin2014 reviewed (approved):
LGTM. I'm much more confident to merge discrete branch after this PR.

jnothman commented:
Further regarding default `n_bins`: showing that 'auto' is useful when not doing a histogram might be hard. I'm +1 for having it more than 2.

TomDLT committed:
address comments

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -140,28 +150,59 @@ def fit(self, X, y=None):
+                bin_edges[jj] = np.asarray(np.percentile(column, quantiles))
+
+            elif self.strategy == 'kmeans':
+                from ..cluster import KMeans  # fixes import loops
```
Please do leave it here. I don't think importing from sklearn.preprocessing should automatically result in importing DBSCAN, etc. Better off keeping the loading as light as possible.

jnothman commented on sklearn/preprocessing/discretization.py:
```diff
@@ -140,28 +150,59 @@ def fit(self, X, y=None):
+                bin_edges[jj] = (centers[1:] + centers[:-1]) * 0.5
+                bin_edges[jj] = np.r_[col_min, bin_edges[jj], col_max]
+
+            if bin_edges[jj][0] == bin_edges[jj][-1] and n_bins[jj] > 2:
```
It's not useful/meaningful to accept n_bins = 1...

qinhanmin2014 replied:
Fine, then changing ``n_bins[jj] > 2`` to ``n_bins[jj] >= 2`` seems enough from my side, or sorry if I'm wrong.

qinhanmin2014 replied:
Also, maybe we need a test for the edge case.

TomDLT committed:
move constant feature check and improve logic

qinhanmin2014 commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -129,23 +143,21 @@ def test_ignored_invalid():
-def test_same_min_max():
+@pytest.mark.parametrize('strategy', ['uniform', 'kmeans', 'quantile'])
+@pytest.mark.parametrize('n_bins', [2, 3])
+def test_same_min_max(strategy, n_bins):
```
Do we test ``n_bins_[0] = 1`` somewhere? Or do you think we don't need such a test?
Also, maybe we don't need to test multiple ``n_bins`` here?

TomDLT replied:
`n_bins = 1` is tested in arrays (`test_invalid_n_bins_array`) and as an integer (`test_invalid_n_bins`).

TomDLT replied:
Oh I misread your comment, I'll add a line to test `n_bins_`.

TomDLT committed:
simpler test

qinhanmin2014 reviewed (approved):
LGTM if CIs are green. ping @jnothman for a final check.

jnothman committed:
Update v0.20.rst
